flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerBase.java

```diff
@@ -0,0 +1,270 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.filemerging;
+
+import org.apache.flink.core.fs.EntropyInjector;
+import org.apache.flink.core.fs.FSDataOutputStream;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.OutputStreamAndPath;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.checkpoint.filemerging.FileMergingCheckpointUtils.SnapshotFileSystemInfo;
+import org.apache.flink.runtime.checkpoint.filemerging.LogicalFile.LogicalFileId;
+import org.apache.flink.runtime.state.CheckpointedStateScope;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+
+/** Base implementation of {@link FileMergingSnapshotManager}. */
+public abstract class FileMergingSnapshotManagerBase implements FileMergingSnapshotManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FileMergingSnapshotManager.class);
+
+    private final String id;
+
+    protected final Executor ioExecutor;
+
+    // file system and directories
+    protected FileSystem fs;
+    protected Path checkpointDir;
+    protected Path sharedStateDir;
+    protected Path taskOwnedStateDir;
+
+    protected int writeBufferSize;
+    private boolean fileSystemInitiated = false;
+
+    protected boolean syncAfterClosingLogicalFile;
+
+    protected PhysicalFile.PhysicalFileDeleter physicalFileDeleter = this::deletePhysicalFile;
+
+    private final Map<SubtaskKey, Path> managedSharedStateDir = new ConcurrentHashMap<>();
+
+    protected Path managedExclusiveStateDir;
+
+    public FileMergingSnapshotManagerBase(String id, Executor ioExecutor) {
+        this.id = id;
+        this.ioExecutor = ioExecutor;
+    }
+
+    @Override
+    public void initFileSystem(SnapshotFileSystemInfo fileSystemInfo) throws IOException {
+        initFileSystem(
+                fileSystemInfo.fs,
+                fileSystemInfo.checkpointBaseDirectory,
+                fileSystemInfo.sharedStateDirectory,
+                fileSystemInfo.taskOwnedStateDirectory);
+    }
+
+    @Override
+    public void addSubtask(SubtaskKey subtaskKey) {
+        try {
+            String managedDirName = subtaskKey.getManagedDirName();
+            Path managedPath = new Path(sharedStateDir, managedDirName);
+            fs.mkdirs(managedPath);
+            managedSharedStateDir.put(subtaskKey, managedPath);
```

Comment:
should we check here if the key is already in the map, to prevent performing mkdir for the same managedPath?

Reply:
Good point. I add de-duplication logic as well as checking if the directory exists in the file system.